src/libmongoc/src/mongoc/mongoc-cluster.c

```diff
@@ -102,7 +102,8 @@ _handle_not_primary_error (mongoc_cluster_t *cluster,
                                           reply,
                                           NULL,
                                           server_stream->sd->max_wire_version,
-                                          server_stream->sd->generation)) {
+                                          server_stream->sd->generation,
+                                          &server_stream->sd->service_id)) {
```

Comment:
`_mongoc_topology_handle_app_error` processes SDAM errors as described in the SDAM specification's [Error Handling](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#error-handling). This now accepts the connection's `service_id` since connection pool clearing is now parameterized by serviceID.